feat: process expired licenses and unlink enterprise learners

muhammad-ammar opened the pull request:
**Description:** Add a management command to iterate through an enterprise's expired licenses and, for each license, call the LMS to unlink the learner from the enterprise.
**JIRA:** https://2u-internal.atlassian.net/browse/ENT-9667


## Testing considerations

- Include instructions for any required manual tests, and any manual testing that has
already been performed.
- Include unit and a11y tests as appropriate
- Consider performance issues.
- Check that Database migrations are backwards-compatible

## Post-review

Squash commits into discrete sets of changes

muhammad-ammar commented on license_manager/apps/subscriptions/management/commands/process_expired_licenses.py:
```diff
@@ -0,0 +1,156 @@
+
+        queryset = License.objects.filter(
+            status__in=[ASSIGNED, ACTIVATED],
+            renewed_to=None,
```
**Note to Reviewers:** I am not sure if `renewed_to=None` should be set or not? I got this from https://github.com/openedx/license-manager/blob/e0d443c035871e451f45b8593d3c68751dd4dc1e/license_manager/apps/subscriptions/models.py#L1876

muhammad-ammar commented on license_manager/apps/subscriptions/management/commands/process_expired_licenses.py:
```diff
@@ -0,0 +1,156 @@
+        expired_subscription_plans = list(
+            SubscriptionPlan.objects.filter(
+                customer_agreement=customer_agreement,
+                expiration_date__lt=now,
```
**Note to Reviewers:** In [expire_subscriptions.py](https://github.com/openedx/license-manager/blob/e0d443c035871e451f45b8593d3c68751dd4dc1e/license_manager/apps/subscriptions/management/commands/expire_subscriptions.py#L131) we are using `expiration_date__lte`. Shouldn't we be using `expiration_date__lt`?

iloveagent57 replied:
Hmm, good question. I *think* from a business perspective, the expiration date configured on subscription plans is exclusive of the active period for the plan. For example, an expiration date of 2025-01-01 means the plan is valid through 2024-12-31 at midnight.

iloveagent57 commented on license_manager/apps/subscriptions/management/commands/process_expired_licenses.py:
```diff
@@ -0,0 +1,156 @@
+
+        logger.info('%s Command started.', log_prefix)
+
+        enterprise_customer_uuids = settings.CUSTOMERS_WITH_EXPIRED_LICENSES_UNLINKING_ENABLED
+        for enterprise_customer_uuid in enterprise_customer_uuids:
+            logger.info('%s Processing started for licenses. Enterprise: [%s]', log_prefix, enterprise_customer_uuid)
+            self.process_expired_licenses(enterprise_customer_uuid, log_prefix, unlink)
+            logger.info('%s Processing completed for licenses. Enterprise: [%s]', log_prefix, enterprise_customer_uuid)
+
+        logger.info('%s Command completed.', log_prefix)
```
Can we make this logic part of the existing `expire_subscriptions` management command?  If we don't, we might be in a situation where we have two crons running around the same time, where `expire_subscriptions` tries to downgrade all the course enrollments for licensed learners to audit mode, and this new `process_expired_licenses` command tries to unlink learners from an enterprise.  If the second one executes first, it might cause the first one to fail.
Inside the [_expire_subscription_plan() method](https://github.com/openedx/license-manager/blob/e0d443c035871e451f45b8593d3c68751dd4dc1e/license_manager/apps/subscriptions/management/commands/expire_subscriptions.py#L73) would be a good place for it.

muhammad-ammar replied:
- So you are saying, we should move the unlink logic into expire_subscriptions management command, correct?
- Would it be ok if I trigger a new async task right after [license_expiration_task](https://github.com/openedx/license-manager/blob/e0d443c035871e451f45b8593d3c68751dd4dc1e/license_manager/apps/subscriptions/management/commands/expire_subscriptions.py#L96-L99)? Any chance that unlinking happens before downgrading the enrollments? OR should I do unlinking without an async task?

iloveagent57 replied:
The expire_subscriptions command currently calls the `license_expiration_task()` synchronously (i.e. without .apply_async()). I would also do the unlinking synchronously from the management command, since we're not in the context of a web request and don't really care if this management command takes several minutes to complete. I would generally only invoke async tasks from a management command if there's lots and lots of work to do *and* the work is inherently parallelizable (for example, when we re-sync data in enterprise-catalog).

iloveagent57 replied:
And yes, correct on your first question.

iloveagent57 commented on license_manager/settings/base.py:
```diff
@@ -477,3 +477,5 @@
 ]
 
 CUSTOMERS_WITH_CUSTOM_LICENSE_EVENTS = ['00000000-1111-2222-3333-444444444444']
+CUSTOMERS_WITH_EXPIRED_LICENSES_UNLINKING_ENABLED = ['76b933cb-bf2a-4c1e-bf44-4e8a58cc37ae']
```
This should be an empty list in base settings.

muhammad-ammar commented on license_manager/apps/subscriptions/management/commands/unlink_expired_licenses.py:
```diff
@@ -0,0 +1,174 @@
+            for prior_renewal in expired_subscription_plan.prior_renewals:
+                expired_subscription_plan_uuids.append(prior_renewal.prior_subscription_plan.uuid)
+
+        queryset = License.objects.filter(
+            status__in=[ASSIGNED, ACTIVATED],
+            renewed_to=None,
+            subscription_plan__uuid__in=expired_subscription_plan_uuids,
+        ).select_related(
```
**Note to Reviewers:** Is the above filtering correct? Specifically asking regarding `status` and `renewed_to`.

iloveagent57 replied:
Yes this is correct.

iloveagent57 commented on license_manager/apps/subscriptions/management/commands/unlink_expired_licenses.py:
```diff
@@ -0,0 +1,174 @@
+            )
+        )
+
+        for expired_subscription_plan in expired_subscription_plans:
+            # exclude subscription plan if there is a renewal
+            if expired_subscription_plan.get_renewal():
+                continue
+
+            expired_subscription_plan_uuids.append(expired_subscription_plan.uuid)
+
+            # include prior renewals
+            for prior_renewal in expired_subscription_plan.prior_renewals:
+                expired_subscription_plan_uuids.append(prior_renewal.prior_subscription_plan.uuid)
```
No you shouldn't need this anymore.

iloveagent57 commented on license_manager/apps/subscriptions/management/commands/unlink_expired_licenses.py:
```diff
@@ -0,0 +1,174 @@
+            for license in licenses:
+                # check if the user associated with the expired license
+                # has any other active licenses with the same customer
+                other_active_licenses = License.for_user_and_customer(
+                    user_email=license.user_email,
+                    lms_user_id=license.lms_user_id,
+                    enterprise_customer_uuid=enterprise_customer_uuid,
+                    active_plans_only=True,
+                    current_plans_only=True,
+                ).exists()
+                if other_active_licenses:
+                    continue
```
👍

macdiesel commented on license_manager/apps/subscriptions/management/commands/unlink_expired_licenses.py:
```diff
@@ -0,0 +1,174 @@
+                    enterprise_customer_uuid,
+                    {
+                        'user_emails': user_emails,
+                        'is_relinkable': False
```
We had a change of requirements from Joe on this. Decision was made here: https://twou.slack.com/archives/C07UVDDRRUZ/p1730470547546569

 `is_relinkable` should be set to true.

macdiesel reviewed (approved):
This looks good. I am going to approve to unblock this work. 

Before merging please ensure that `is_relinkable` is set to true when unlinking these learners.

muhammad-ammar commented:
@iloveagent57 @macdiesel Thank you for the review. I have addressed your feedback and also added unit tests. I will do some more manual testing and then merge.
